### DISCUSSION TITLE
Make diesel pure rust with no external depedencies

### DISCUSSION THREAD

**Discussion by AlbertMarashi** [diesel-rs/diesel#3480] [Ideas]
Having to add external dependencies and involving them in the linking process adds complication to the compilation process and I think it would be nice to have `postgres` support with just pure rust-based code.

There seems to be a crate that might be able to serve this functionality
https://github.com/sfackler/rust-postgres

Any reason why there hasn't been migration towards a pure rust-based system?

**Comment by weiznich** [comment]
Generally speaking: That's not possible, because the  `sqlite` backend requires by definition linking `libsqlite3`.

For postgres that request is a duplicate of https://github.com/diesel-rs/diesel/issues/2084, https://github.com/diesel-rs/diesel/pull/2375 and https://github.com/diesel-rs/diesel/pull/2257

The summary of these issues is: We are activly looking for someone that wants to develop and maintain such a connection as third party crate. We as diesel team do not have the capacity for that.

**Reply by AlbertMarashi** [reply]
@weiznich My team and I are open to potentially developing and maintaining such a crate. Is there a discord where I could ask questions should I run into any issues?

**Reply by weiznich** [reply]
We do not run a discord server for diesel. We do maintain an [gitter room](https://gitter.im/diesel-rs/diesel). Otherwise you can ask question here in the discussion forum. (That's better suited for more complex discussions)

**Comment by AlbertMarashi** [comment]
Just reposting this from #3482 since it should appropriately belong here.

It would be nice if diesel could move towards native rust-based [postgres](https://docs.rs/postgres/latest/postgres/index.html) clients. Implementing it doesn't seem too difficult and I think it would only involve implementing the [Connection](https://docs.rs/diesel/latest/diesel/connection/trait.Connection.html) trait and some other stuff to get it working.

~~Async could also be supported through [tokio-postgres](https://docs.rs/tokio-postgres/latest/tokio_postgres/) in a similar way to [diesel-async](https://crates.io/crates/diesel-async)~~

It appears that `diesel-async` uses `tokio-postgres` under the hood. This should really be noted in the documentation of `diesel` to let everyone know that it is an option, as I almost didn't use diesel because it didn't have async support
1. there is an async integration for diesel
2. it doesn't appear to depend on any external `libpq` libraries

**Reply by weiznich** [reply]
> It appears that diesel-async uses tokio-postgres under the hood. This should really be noted in the documentation of diesel to let everyone know that it is an option, as I almost didn't use diesel because it didn't have async support
>
> *    there is an async integration for diesel
> *    it doesn't appear to depend on any external libpq libraries

This is already documented on the diesel web side under "Community Projects". That's the correct location for linking this crate, as to repeat that again: `diesel-async` is neither part of diesel itself, nor is it maintained by the diesel organization, nor does it currently provide the same guarantees around stability and support. In fact I do consider that crate as experimental in the current state. To change that I would need to spend quite a lot of time to figure out possible workarounds for language level issues, which is not something I currently have the capacity to do while maintaining diesel itself.
